agent/src/main/java/com/newrelic/agent/android/NewRelic.java

```diff
@@ -927,13 +927,29 @@ public static boolean removeAllAttributes() {
      * Sets a user ID attribute.
      *
      * @param userId The user ID as string value
-     * @return true if successful, false if the operation did not complete as anticipated.
+     * @return true if userId attribute as created or updated.
      */
     public static boolean setUserId(String userId) {
         StatsEngine.notice().inc(MetricNames.SUPPORTABILITY_API
                 .replace(MetricNames.TAG_NAME, "setUserId"));
 
-        return AnalyticsControllerImpl.getInstance().setAttribute(AnalyticsAttribute.USER_ID_ATTRIBUTE, userId);
+        final AnalyticsControllerImpl controller = AnalyticsControllerImpl.getInstance();
+        final AnalyticsAttribute userIdAttr = controller.getAttribute(AnalyticsAttribute.USER_ID_ATTRIBUTE);
+
+        if (userIdAttr != null) {
+            if (userId == null || userId.isEmpty()) {
+                controller.removeAttribute(AnalyticsAttribute.USER_ID_ATTRIBUTE);
+            }
+
+            if (userIdAttr.getStringValue() != userId) {
+                Harvest.harvestNow(true, false);   // call non-blocking harvest
```

Comment:
Blocking methods should never be called from the NewRelic API level. The new 2nd arg says whether to wait on harvest completion, which is no in  this case.